repo/repo.go

```diff
@@ -1,282 +1,331 @@
+type BlobOperation byte
 
-	return "."
-}
+const (
+	BlobRead   = 'R' // A blob has been read
+	BlobWrite  = 'W' // A blob has been written
+	BlobDelete = 'D' // A blob has been deleted
+)
 
-// getPath returns the path for a file type in the repo.
-func (s *Server) getPath(r *http.Request, fileType string) (string, error) {
-	if !s.isValidType(fileType) {
-		return "", errors.New("invalid file type")
+// BlobMetricFunc is the callback signature for blob metrics. Such a callback
+// can be passed in the Options to keep track of various metrics.
+// objectType: one of ObjectTypes
+// operation: one of the BlobOperations above
+// nBytes: the number of bytes affected, or 0 if not relevant
+type BlobMetricFunc func(objectType string, operation BlobOperation, nBytes uint64)
+
+// ServeHTTP performs strict matching on the repo part of the URL path and
+// dispatches the request to the appropriate handler.
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	urlPath := r.URL.Path
+	if urlPath == "/" {
+		// TODO: add HEAD and GET
+		switch r.Method {
+		case "POST":
+			h.createRepo(w, r)
+		default:
+			httpMethodNotAllowed(w, []string{"POST"})
+		}
+		return
+	} else if urlPath == "/config" {
+		switch r.Method {
+		case "HEAD":
+			h.checkConfig(w, r)
+		case "GET":
+			h.getConfig(w, r)
+		case "POST":
+			h.saveConfig(w, r)
+		case "DELETE":
+			h.deleteConfig(w, r)
+		default:
+			httpMethodNotAllowed(w, []string{"HEAD", "GET", "POST", "DELETE"})
+		}
+		return
+	} else if objectType, objectID := h.getObject(urlPath); objectType != "" {
+		if objectID == "" {
+			// TODO: add HEAD
+			switch r.Method {
+			case "GET":
+				h.listBlobs(w, r)
+			default:
+				httpMethodNotAllowed(w, []string{"GET"})
+			}
+			return
+		} else {
+			switch r.Method {
+			case "HEAD":
+				h.checkBlob(w, r)
+			case "GET":
+				h.getBlob(w, r)
+			case "POST":
+				h.saveBlob(w, r)
+			case "DELETE":
+				h.deleteBlob(w, r)
+			default:
+				httpMethodNotAllowed(w, []string{"HEAD", "GET", "POST", "DELETE"})
+			}
+			return
+		}
 	}
-	return join(s.Path, s.getRepo(r), fileType)
+	httpDefaultError(w, http.StatusNotFound)
 }
 
-// getFilePath returns the path for a file in the repo.
-func (s *Server) getFilePath(r *http.Request, fileType, name string) (string, error) {
-	if !s.isValidType(fileType) {
-		return "", errors.New("invalid file type")
+// getObject parses the URL path and returns the objectType and objectID,
+// if any. The objectID is optional.
+func (h *Handler) getObject(urlPath string) (objectType, objectID string) {
+	if m := BlobPathRE.FindStringSubmatch(urlPath); len(m) > 0 {
+		if len(m) == 2 || m[2] == "" {
+			return m[1], ""
+		}
+		return m[1], m[2]
+	} else {
+		return "", ""
 	}
+}
 
-	if s.isHashed(fileType) {
-		if len(name) < 2 {
-			return "", errors.New("file name is too short")
-		}
+// getSubPath returns the path for a file or subdir in the root of the repo.
+func (h *Handler) getSubPath(name string) string {
```

Comment:
I'm not sure this warrants a function. Maybe just inline it?

Reply:
I forgot the reason, but with the current contents it indeed does not add much value. Will change, unless I recall a good reason. :)